File: protect/src/main/java/com/niu/protect/lib/receiver/BroadcastManager.java
package com.niu.protect.lib.receiver;

import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;

import com.niu.protect.core.Constants;
import com.niu.protect.tools.ILog;
import java.util.List;
public class BroadcastManager {
    static AppInstallReceiver mAppInstallReceiver;
    static ShutDownReciver mShutDownReciver;
    static ScreenReceiver screenReceiver;

    /**
     * Android 14(targetSdk 34)起，动态注册非「纯系统广播」的 receiver 必须显式声明
     * RECEIVER_EXPORTED / NOT_EXPORTED，否则抛 SecurityException。这些 receiver 只接收
     * 系统广播和本应用自发广播，用 NOT_EXPORTED 即可。
     * 漏了这个标志会导致无障碍服务 onCreate 崩溃、服务永远绑不上。
     */
    private static void registerReceiverCompat(Context context, android.content.BroadcastReceiver receiver, IntentFilter filter) {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
            context.registerReceiver(receiver, filter, Context.RECEIVER_NOT_EXPORTED);
        } else {
            context.registerReceiver(receiver, filter);
        }
    }

    public static void startScreenBroadcastReceiver(Context context) {
        screenReceiver = new ScreenReceiver();
        IntentFilter filter = new IntentFilter();
        filter.addAction("android.intent.action.SCREEN_ON");
        filter.addAction("android.intent.action.SCREEN_OFF");
        filter.addAction("android.intent.action.USER_PRESENT");
        filter.addAction(Constants.ACTION_ACCESSIBILITY_START);
        filter.addAction(Constants.ACTION_ACCESSIBILITY_STOP);
        filter.addAction("android.net.conn.CONNECTIVITY_CHANGE");
        registerReceiverCompat(context, screenReceiver, filter);
    }

    public static void registerShutDownReciver(Context context) {
        ILog.d("BroadcastManager", "注册广播--");
        if (mShutDownReciver != null) {
            return;
        }
        mShutDownReciver = new ShutDownReciver();
        IntentFilter filter = new IntentFilter();
        filter.addAction("android.intent.action.ACTION_SHUTDOWN");
        filter.addAction("android.intent.action.MEDIA_MOUNTED");
        filter.addAction("android.intent.action.MEDIA_UNMOUNTED");
        registerReceiverCompat(context, mShutDownReciver, filter);
        registerAppInstallReceiver(context);
        startScreenBroadcastReceiver(context);
    }

    public static void registerAppInstallReceiver(Context context) {
        if (mAppInstallReceiver != null) {
            return;
        }
        mAppInstallReceiver = new AppInstallReceiver();
        IntentFilter filter = new IntentFilter();
        filter.addAction("android.intent.action.INSTALL_PACKAGE");
        filter.addAction("android.intent.action.PACKAGE_ADDED");
        filter.addAction("android.intent.action.PACKAGE_REMOVED");
        filter.addDataScheme("package");
        registerReceiverCompat(context, mAppInstallReceiver, filter);
    }

    public static void unRegisterShutDownReciver(Context context) {
        ILog.d("BroadcastManager", "注册广播-unRegisterShutDownReciver-");
        ShutDownReciver shutDownReciver = mShutDownReciver;
        if (shutDownReciver != null) {
            context.unregisterReceiver(shutDownReciver);
            mShutDownReciver = null;
        }
        unAppInstallReceiver(context);
        unScreenBroadcastReceiver(context);
    }

    public static void unAppInstallReceiver(Context context) {
        AppInstallReceiver appInstallReceiver = mAppInstallReceiver;
        if (appInstallReceiver != null) {
            context.unregisterReceiver(appInstallReceiver);
            mAppInstallReceiver = null;
        }
    }

    public static void sendAccessibilityStart(Context context) {
        Intent intent = new Intent(Constants.ACTION_ACCESSIBILITY_START);
        context.sendBroadcast(intent);
    }

    public static void sendAccessibilityStop(Context context) {
        Intent intent = new Intent(Constants.ACTION_ACCESSIBILITY_STOP);
        context.sendBroadcast(intent);
    }

    public static void unScreenBroadcastReceiver(Context context) {
        ScreenReceiver screenReceiver2 = screenReceiver;
        if (screenReceiver2 != null) {
            context.unregisterReceiver(screenReceiver2);
        }
    }

    private void searchRigisterBroadcast(Context context) {
        Intent intent = new Intent();
        intent.setAction("android.intent.action.ACTION_SHUTDOWN");
        PackageManager pm = context.getPackageManager();
        List<ResolveInfo> resolveInfos = pm.queryBroadcastReceivers(intent, 0);
        if (resolveInfos != null) {
            resolveInfos.isEmpty();
        }
    }
}

File: corelib/src/main/java/im/niu/corelib/manager/BroadcastManager.kt
package im.niu.corelib.manager

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import im.niu.corelib.receiver.AppReceiver
import im.niu.corelib.receiver.MainReceiver
import im.niu.corelib.utils.ILog

class BroadcastManager {



    companion object {
        var mAppInstallReceiver: AppReceiver = AppReceiver()
        var mainReceiver: MainReceiver = MainReceiver()
        private val TAG: String = "BroadCast"

        private fun searchRigisterBroadcast(context: Context,action:String): Boolean {
            val intent = Intent()
            intent.action = action
            val pm: PackageManager = context.packageManager
            val resolveInfos: List<ResolveInfo> = pm.queryBroadcastReceivers(intent, 0)
            for (resolveInfo in resolveInfos){
                ILog.d(TAG, resolveInfo.toString())
                if (resolveInfo.activityInfo.packageName == context.packageName){
                    return true
                }
            }

            return false
        }
        /**
         * 注册一个广播接收器来监听系统相关的广播事件。
         * 本函数用于设置广播接收器所关注的事件类型，并将其注册到系统中。
         *
         * @param context 应用程序的上下文对象，用于注册广播接收器。
         */
        fun register(context: Context) {
            var filter = IntentFilter()
            if(!searchRigisterBroadcast(context,Intent.ACTION_BOOT_COMPLETED)){
                // 注册屏幕开关相关的广播动作
                filter.addAction("android.intent.action.SCREEN_ON")
                filter.addAction("android.intent.action.SCREEN_OFF")
                filter.addAction("android.intent.action.USER_PRESENT") // 注册用户解锁广播动作

                // 注册网络连接状态改变的广播动作
                filter.addAction("android.net.conn.CONNECTIVITY_CHANGE")
                // 注册系统关机和媒体设备挂载、卸载的广播动作
                filter.addAction("android.intent.action.ACTION_SHUTDOWN")
                filter.addAction("android.intent.action.MEDIA_MOUNTED")
                filter.addAction("android.intent.action.MEDIA_UNMOUNTED")
                context.registerReceiver(mainReceiver, filter)
                filter = IntentFilter()
            }
            if(!searchRigisterBroadcast(context,Intent.ACTION_PACKAGE_ADDED)) {
                // 创建一个IntentFilter用于过滤广播
                // 注册与软件包安装、删除相关的广播动作
                filter.addAction("android.intent.action.INSTALL_PACKAGE")
                filter.addAction("android.intent.action.PACKAGE_ADDED")
                filter.addAction("android.intent.action.PACKAGE_REMOVED")
                filter.addDataScheme("package") // 添加对软件包名称的过滤
                // 使用上下文对象注册广播接收器，并应用之前设置的IntentFilter
                context.registerReceiver(mAppInstallReceiver, filter)
            }
        }

        fun unRegister(context: Context){
            context.unregisterReceiver(mainReceiver)
            context.unregisterReceiver(mAppInstallReceiver)
        }
    }
}